Communications/PacketBuffer.cs

```diff
@@ -0,0 +1,373 @@
+        /// <returns> Next packet if buffer is not empty, or null otherwise. </returns>
+        public override Packet Peek()
+        {
+            if (this.PacketCache == null) { this.PacketCache = Dequeue(); } // Store next packet into cache so it can be removed on next call of dequeue
+            return this.PacketCache;
+        }
+
+        /// <summary> Get next packet and remove it. </summary>
+        /// <remarks> It will return null if the buffer is empty instead of throwing an exception. </remarks>
+        /// <returns> Next packet if buffer is not empty, or null otherwise. </returns>
+        public override Packet Dequeue()
+        {
+            // Use the packet stored in cache if available
+            if (this.PacketCache != null)
+            {
+                Packet LastPacket = this.PacketCache;
+                this.PacketCache = null;
+                return LastPacket;
+            }
+
+            // Find next packet
+            if (this.TokenBuckets[this.CurrentBucket] < 0 || this.Buffers[this.CurrentBucket].Peek() == null)
+            {
+                if (NextBucket()) { return null; } // Can't find any packet
+            }
+
+            // Update token bucket
+            Packet Next = this.Buffers[this.CurrentBucket].Dequeue();
+            this.TokenBuckets[this.CurrentBucket] -= Next.GetLength();
+
+            // Refill token bucket if it is not enough
+            if (this.TokenBuckets[this.CurrentBucket] < 0)
+            {
+                AddToken(1 + (-this.TokenBuckets[this.CurrentBucket]) / this.BandwidthAllocations[this.CurrentBucket]);
+
+                // Switch to the buffer of next prioirity
+                this.CurrentBucket++;
+                this.CurrentBucket = this.CurrentBucket >= this.NBuffers ? 0 : this.CurrentBucket;
+            }
+            return Next;
+        }
+
+        /// <summary> Get the total number of packets in the buffer. </summary>
+        public override int Count
+        {
+            get
+            {
+                int Sum = 0;
+                foreach (PacketBuffer Buffer in this.Buffers) { Sum += Buffer.Count; }
+                return Sum;
+            }
+        }
+    }
+
+    /// <summary>
+    /// This defines a generic controller.
+    /// The order of absolute priority is EMERGENT > HIGH == MEDIUM == LOW > LOWEST,
+    /// and the bandwidth allocation of HIGH : MEDIUM : LOW is 3 : 2 : 1.
+    /// </summary>
+    public class GenericController : PacketBuffer
+    {
+        public readonly QueueBuffer[] Buffers;
+        public readonly PriorityBuffer PriorityBuffer;
+        public readonly BandwidthControlBuffer BandwidthBuffer;
+
+        /// <summary> Construct a generic Controller. </summary>
+        public GenericController()
+        {
+            this.Buffers = new QueueBuffer[5];
+            for (int i = 0; i < 5; i++) { this.Buffers[i] = new QueueBuffer(); }
```

Comment:
Would it make more sense to use say `this.Buffers.Count` instead of hard-coding `5` here?